src/System.Text.Primitives/System/Text/Encoding/Utf8/Utf8Encoder.cs

```diff
@@ -298,120 +298,110 @@ internal static bool TryComputeEncodedBytes(ReadOnlySpan<uint> utf32, out int by
+                            goto non_ascii;
+                        }
+                        if (bytesWrittenTemp > outputBytesAvailable - 1)
+                        {
+                            goto need_more;
+                        }
+                        *(bytesWrittenTemp + pUtf8) = (byte)codePoint;
+                        bytesWrittenTemp++;
+                    }
 
-                            if (codePoint > Utf8OneByteLastCodePoint)
-                            {
-                                goto non_ascii;
-                            }
-                            if (bytesWrittenTemp > outputBytesAvailable - 1)
+                    non_ascii:
+                    for (; i < inputCharacters; i++)
+                    {
+                        codePoint = *(pUtf16 + i);
+                        if (codePoint <= Utf8OneByteLastCodePoint)
+                        {
+                            goto ascii;
+                        }
+                        else if (codePoint <= Utf8TwoBytesLastCodePoint)
+                        {
+                            if (bytesWrittenTemp > outputBytesAvailable - 2)
                             {
                                 goto need_more;
                             }
-                            *(pUtf8 + bytesWrittenTemp) = (byte)codePoint;
+                            *(bytesWrittenTemp + pUtf8) = (byte)(((codePoint >> 6) & b0001_1111U) | b1100_0000U);
+                            bytesWrittenTemp++;
+                            *(bytesWrittenTemp + pUtf8) = (byte)(((codePoint) & b0011_1111U) | b1000_0000U);
                             bytesWrittenTemp++;
                         }
-
-                    non_ascii:
-                        for (; i < inputCharacters; i++)
+                        else if (codePoint >= Utf16HighSurrogateFirstCodePoint && codePoint <= Utf16HighSurrogateLastCodePoint)
                         {
-                            codePoint = *(pUtf16 + i);
-                            if (codePoint <= Utf8OneByteLastCodePoint)
+                            if ((bytesWrittenTemp > outputBytesAvailable - 4) || (++i >= inputCharacters))
                             {
-                                goto ascii;
-                            }
-                            else if (codePoint <= Utf8TwoBytesLastCodePoint)
-                            {
-                                if (bytesWrittenTemp > outputBytesAvailable - 2)
-                                {
-                                    goto need_more;
-                                }
-                                *(pUtf8 + bytesWrittenTemp) = (byte)(((codePoint >> 6) & b0001_1111U) | b1100_0000U);
-                                bytesWrittenTemp++;
-                                *(pUtf8 + bytesWrittenTemp) = (byte)(((codePoint) & b0011_1111U) | b1000_0000U);
-                                bytesWrittenTemp++;
+                                goto need_more;
```

Comment:
I do not think that the charactersConsumed will be always right if you take this path - even if `++i >= inputCharacters` is false, `i` is still going to be incremented. Make sure to add the test for this... .